core/module_specifier.rs

```diff
@@ -80,3 +109,169 @@ impl PartialEq<String> for ModuleSpecifier {
+
+      assert_eq!(
+        ModuleSpecifier::resolve_from_cwd(r"./tests/006_url_imports.ts")
+          .unwrap()
+          .to_string(),
+        expected_url
+      );
+      assert_eq!(
+        ModuleSpecifier::resolve_from_cwd(r".\tests\006_url_imports.ts")
+          .unwrap()
+          .to_string(),
+        expected_url
+      );
+    } else {
+      assert_eq!(
+        ModuleSpecifier::resolve_from_cwd("./tests/006_url_imports.ts")
+          .unwrap()
+          .to_string(),
+        format!(
+          "file://{}{}",
+          cwd_string.as_str(),
+          "tests/006_url_imports.ts"
+        ),
+      );
+    }
+
+    assert_eq!(
+      ModuleSpecifier::resolve_from_cwd(
+        "http://deno.land/core/tests/006_url_imports.ts"
+      ).unwrap()
+      .to_string(),
+      "http://deno.land/core/tests/006_url_imports.ts",
+    );
+    assert_eq!(
+      ModuleSpecifier::resolve_from_cwd(
+        "https://deno.land/core/tests/006_url_imports.ts"
+      ).unwrap()
+      .to_string(),
+      "https://deno.land/core/tests/006_url_imports.ts",
+    );
+  }
+
+  #[test]
+  fn test_resolve() {
+    assert_eq!(
+      ModuleSpecifier::resolve(
+        "./005_more_imports.ts",
+        "http://deno.land/core/tests/006_url_imports.ts",
+      ).unwrap()
+      .to_string(),
+      "http://deno.land/core/tests/005_more_imports.ts"
+    );
+
+    assert_eq!(
+      ModuleSpecifier::resolve(
+        "../005_more_imports.ts",
+        "http://deno.land/core/tests/006_url_imports.ts",
+      ).unwrap()
+      .to_string(),
+      "http://deno.land/core/005_more_imports.ts"
+    );
+
+    // absolute URL
+    assert_eq!(
+      ModuleSpecifier::resolve(
+        "http://deno.land/core/tests/005_more_imports.ts",
+        "http://deno.land/core/tests/006_url_imports.ts",
+      ).unwrap()
+      .to_string(),
+      "http://deno.land/core/tests/005_more_imports.ts",
+    );
+
+    // absolute path
+    if cfg!(target_os = "windows") {
+      assert_eq!(
+        ModuleSpecifier::resolve(
+          r"C:/deno/tests/005_more_imports.ts",
+          r"C:/deno/tests/006_url_imports.ts",
+        ).unwrap()
+        .to_string(),
+        "file:///C:/deno/tests/005_more_imports.ts",
+      );
+      assert_eq!(
+        ModuleSpecifier::resolve(
+          r"C:\deno\tests\005_more_imports.ts",
+          r"C:/deno/tests/006_url_imports.ts",
+        ).unwrap()
+        .to_string(),
+        "file:///C:/deno/tests/005_more_imports.ts",
+      );
+    } else {
+      assert_eq!(
+        ModuleSpecifier::resolve(
+          "/dev/core/tests/005_more_imports.ts",
+          "/dev/core/tests/006_url_imports.ts",
+        ).unwrap()
+        .to_string(),
+        "file:///dev/core/tests/005_more_imports.ts",
+      );
+    }
+  }
+
+  #[test]
+  fn test_resolve_bad_specifier() {
+    assert!(
+      ModuleSpecifier::resolve(
+        "005_more_imports.ts",
+        "http://deno.land/core/tests/006_url_imports.ts",
+      ).is_err()
```

Comment:
Why is this an error? I'm confused...

Reply:
> If specifier does not start with the character U+002F SOLIDUS (/), the two-character sequence U+002E FULL STOP, U+002F SOLIDUS (./), or the three-character sequence U+002E FULL STOP, U+002E FULL STOP, U+002F SOLIDUS (../), return failure.

Reply:
Currently we return `url::ParseError::RelativeUrlWithCannotBeABaseBase` for this case - it is really non-descriptive an causes a lot of confusion (see #2338). 

How about we add `ModuleSpecifierError` which wraps `url::ParseError` and extends it with one option that is precisely designated to handle this case?

Reply:
I have a patch that does exactly that (https://github.com/denoland/deno/compare/master...piscisaureus:err2?expand=1#diff-28c0e34cfc9b15337fd837fc562cae19).
It also does a bunch of other junk that didn't work out and I'm going to throw away, but it may be a good starting point.

Reply:
@piscisaureus that's nice! I just [pushed my minimal implementation](https://github.com/denoland/deno/pull/2572/commits/e422967ef15118c2875ab4f4969fcad25abfb0e3) for `ModuleSpecifierError` but yours looks much more idiomatic and better. I'm gonna apply it